File: MPChartLib/src/main/java/com/github/mikephil/charting/data/Entry.java

package com.github.mikephil.charting.data;

import android.os.Parcel;
import android.os.ParcelFormatException;
import android.os.Parcelable;

/**
 * Class representing one entry in the chart. Might contain multiple values.
 * Might only contain a single yValue depending on the used constructor.
 * 
 * @author Philipp Jahoda
 */
public class Entry implements Parcelable {

    /** the yPx yValue */
    private float y = 0f;

    /** the xPx yValue */
    private float x = 0f;

    /** optional spot for additional data this Entry represents */
    private Object mData = null;

    /**
     * A Entry represents one single entry in the chart.
     *
     * @param x the xPx yValue
     * @param y the yPx yValue (the actual yValue of the entry)
     */
    public Entry(float x, float y) {
        this.y = y;
        this.x = x;
    }

    /**
     * A Entry represents one single entry in the chart.
     *
     * @param x the xPx yValue
     * @param y the yPx yValue (the actual yValue of the entry)
     * @param data Spot for additional data this Entry represents.
     */
    public Entry(float x, float y, Object data) {
        this(x, y);
        this.mData = data;
    }

    /**
     * Returns the xPx-yValue of this Entry object.
     * 
     * @return
     */
    public float getX() {
        return x;
    }

    /**
     * Sets the xPx-yValue of this Entry object.
     * 
     * @param x
     */
    public void setX(float x) {
        this.x = x;
    }

    /**
     * Returns the yPx yValue of this Entry.
     * 
     * @return
     */
    public float getY() {
        return y;
    }

    /**
     * Sets the yPx-yValue for the Entry.
     * 
     * @param y
     */
    public void setY(float y) {
        this.y = y;
    }

    /**
     * Returns the data, additional information that this Entry represents, or
     * null, if no data has been specified.
     * 
     * @return
     */
    public Object getData() {
        return mData;
    }

    /**
     * Sets additional data this Entry should represent.
     * 
     * @param data
     */
    public void setData(Object data) {
        this.mData = data;
    }

    /**
     * returns an exact copy of the entry
     * 
     * @return
     */
    public Entry copy() {
        Entry e = new Entry(x, y, mData);
        return e;
    }

    /**
     * Compares yValue, xIndex and data of the entries. Returns true if entries
     * are equal in those points, false if not. Does not check by hash-code like
     * it's done by the "equals" method.
     * 
     * @param e
     * @return
     */
    public boolean equalTo(Entry e) {

        if (e == null)
            return false;

        if (e.mData != this.mData)
            return false;

        if (Math.abs(e.x - this.x) > 0.000001f)
            return false;

        if (Math.abs(e.y - this.y) > 0.000001f)
            return false;

        return true;
    }

    /**
     * returns a string representation of the entry containing xPx-index and yValue
     */
    @Override
    public String toString() {
        return "Entry, x: " + x + " y (sum): " + getY();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeFloat(this.x);
        dest.writeFloat(this.y);
        if (mData != null) {
            if (mData instanceof Parcelable) {
                dest.writeInt(1);
                dest.writeParcelable((Parcelable) this.mData, flags);
            } else {
                throw new ParcelFormatException("Cannot parcel an Entry with non-parcelable data");
            }
        } else {
            dest.writeInt(0);
        }
    }

    protected Entry(Parcel in) {
        this.x = in.readFloat();
        this.y = in.readFloat();
        if (in.readInt() == 1) {
            this.mData = in.readParcelable(Object.class.getClassLoader());
        }
    }

    public static final Parcelable.Creator<Entry> CREATOR = new Parcelable.Creator<Entry>() {
        public Entry createFromParcel(Parcel source) {
            return new Entry(source);
        }

        public Entry[] newArray(int size) {
            return new Entry[size];
        }
    };
}

File: MPChartLib/src/main/java/com/github/mikephil/charting/data/Entry.kt
package com.github.mikephil.charting.data

import android.graphics.drawable.Drawable
import android.os.Parcel
import android.os.ParcelFormatException
import android.os.Parcelable
import com.github.mikephil.charting.utils.Utils
import java.io.Serializable
import kotlin.math.abs

/**
 * Class representing one entry in the chart. Might contain multiple values.
 * Might only contain a single value depending on the used constructor.
 */
open class Entry : BaseEntry, Parcelable, Serializable {

    private var _x: Float = 0f
    open var x: Float
        get() = _x
        set(value) {
            _x = value
        }

    constructor()

    /**
     * A Entry represents one single entry in the chart.
     *
     * @param x the x value
     * @param y the y value (the actual value of the entry)
     */
    constructor(x: Float, y: Float) : super(y) {
        this._x = x
    }

    /**
     * A Entry represents one single entry in the chart.
     *
     * @param x    the x value
     * @param y    the y value (the actual value of the entry)
     * @param data Spot for additional data this Entry represents.
     */
    constructor(x: Float, y: Float, data: Any?) : super(y, data) {
        this._x = x
    }

    /**
     * A Entry represents one single entry in the chart.
     *
     * @param x the x value
     * @param y the y value (the actual value of the entry)
     * @param icon icon image
     */
    constructor(x: Float, y: Float, icon: Drawable?) : super(y, icon) {
        this._x = x
    }

    /**
     * A Entry represents one single entry in the chart.
     *
     * @param x the x value
     * @param y the y value (the actual value of the entry)
     * @param icon icon image
     * @param data Spot for additional data this Entry represents.
     */
    constructor(x: Float, y: Float, icon: Drawable?, data: Any?) : super(y, icon, data) {
        this._x = x
    }

    /**
     * returns an exact copy of the entry
     *
     * @return
     */
    open fun copy(): Entry? {
        val e = Entry(x, y, data)
        return e
    }

    /**
     * Compares value, xIndex and data of the entries. Returns true if entries
     * are equal in those points, false if not. Does not check by hash-code like
     * it's done by the "equals" method.
     *
     * @param e
     * @return
     */
    fun equalTo(e: Entry?): Boolean {
        if (e == null) return false

        if (e.data !== this.data) return false

        if (abs((e.x - this.x).toDouble()) > Utils.FLOAT_EPSILON) return false

        if (abs((e.y - this.y).toDouble()) > Utils.FLOAT_EPSILON) return false

        return true
    }

    /**
     * returns a string representation of the entry containing x-index and value
     */
    override fun toString(): String {
        return "Entry x=$x y=$y"
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeFloat(this.x)
        dest.writeFloat(this.y)
        if (data != null) {
            if (data is Parcelable) {
                dest.writeInt(1)
                dest.writeParcelable(data as Parcelable?, flags)
            } else {
                throw ParcelFormatException("Cannot parcel an Entry with non-parcelable data")
            }
        } else {
            dest.writeInt(0)
        }
    }

    protected constructor(`in`: Parcel) {
        this._x = `in`.readFloat()
        this._y = `in`.readFloat()
        if (`in`.readInt() == 1) {
            this.data = `in`.readParcelable(Any::class.java.classLoader)
        }
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<Entry?> = object : Parcelable.Creator<Entry?> {
            override fun createFromParcel(source: Parcel): Entry {
                return Entry(source)
            }

            override fun newArray(size: Int): Array<Entry?> {
                return arrayOfNulls(size)
            }
        }
    }
}
